### PR TITLE
Trait [Info] LINQ-like Access Changes

### PR DESCRIPTION
**Pull request by atlimit8**
_TODO: rewrite_

### CODE REVIEW

**Comment by pchote** [comment]
Can we please qualify the ActorInfo queries with `TraitInfo`?
`actorInfo.HasTraitInfo<>`, `actorInfo.TraitInfo<>`, `actorInfo.TraitInfoOrDefault<>`

ActorInfo is (slightly) more than just a collection of TraitInfos, so it's a bit weird not to qualify this further.

**Comment by pchote** [comment]
Another idea could be to add more linq-like named queries to the TraitInfo dictionary: `actorInfo.Traits.Contains<FooInfo>()`, `actorInfo.Traits.First<FooInfo>()`, `actorInfo.Traits.FirstOrDefault<FooInfo>()`, `actorInfo.Traits.ManyOrDefault<FooInfo>()`.
I think I prefer this to my earlier suggestion above.

**Comment by atlimit8** [comment]
How about *info*`.[Has]Trait[OrDefault]` and *actor*`.[Has]TraitInfo[OrDefault]`?

**Comment by pchote** [comment]
Perhaps `.FirstTrait[Info]<>`, `.FirstTrait[Info]OrDefault<>`, `.Trait[Info]s<>`?

Your future prs that increase the strictness around dependencies could then also have `.SingleTrait[Info]<>`.

**Comment by pchote** [comment]
Another though / constraint to consider: we can make this even more consistent with linq by allowing a filter function to be passed to the methods that make sense.  We could then replace e.g.:
```
self.TraitsImplementing<Turreted>().FirstOrDefault(t => t.Name == info.Turret)
```
with
```
self.FirstTraitOrDefault<Turreted>(t => t.Name == info.Turret)

**Comment by atlimit8** [comment]
Already doing it.

**Comment by atlimit8** [comment]
It has the added bonus of avoiding the creation unnecessary `IEnumerable`s.

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by atlimit8** [comment]
WIP now

**Comment by atlimit8** [comment]
Rebased and more or less finished (with part1).

**Comment by RoosterDragon** [comment]
What are the motivations behind providing custom enumerators/methods for Where-type queries and a custom ToArray implementation?

**Comment by atlimit8** [comment]
@RoosterDragon, hopefully less state/overhead for commonly used patterns. As I mentioned on the IRC channel, it would be nice to see how this affects performance. The trait to trait info lookup changes may reduce performance; Hopefully, refactoring `TraitDictionary` for use by both `Actor` for traits and `ActorInfo` for trait info will help.

**Comment by RoosterDragon** [comment]
I would suggest dropping the custom Where enumerators. If there's a perf problem from a Where the much better solution would be to use a plain old loop which skips all possible overhead of enumerators in the first place. The overloads that take predicates are useful, but I would just implement them in terms of the existing LINQ operators for simplicity.

I don't think the complexity of the custom ToArray justifies the gains - we don't seem to call it often enough unless you can point me at a common use case that I'm not familiar with (I don't poke around traits too much so I could easily have missed something). Typically I would imagine calling the usual LINQ ToArray would be good enough.

File: OpenRA.Game/Actor.cs [hunk 1]
```diff
@@ -209,22 +215,67 @@ public override string ToString()
...
+		{
+			return World.TraitDict.FirstOrDefault<T>(this);
+		}
+
+		public T FirstTraitOrDefault<T>(Func<T, bool> predicate)
+		{
+			return World.TraitDict.FirstOrDefault<T>(this, predicate);
+		}
+
+		public bool TraitsAny<T>(Func<T, bool> predicate)
```
Comment (RoosterDragon) [inline OpenRA.Game/Actor.cs]: Maybe ```AnyTrait``` ?

**Comment by atlimit8** [comment]
I might be able to use `ForEachActorHavingTrait<T>([Func<actor, T, bool> predicate], Action<Actor> body)`, `CountActorsHavingTrait<T>([Func<actor, T, bool> predicate])`, `ForEachActorTrait<T>(Action<Actor, T> body)` instead of  `ActorsWithTrait([Func<actor, T, bool> predicate])`
``` C#
PlayerResources.Tick(Actor self) :
   ResourceCapacity = self.World.ActorsWithTrait<IStoreResources>((a, t) => a.Owner == owner)..Sum(a => a.Trait.Capacity);
---
OpenRA.Game/World.cs-280-                               using (new PerfSample("tick_idle"))
OpenRA.Game/World.cs:281:                                       foreach (var ni in ActorsWithTrait<INotifyIdle>())
OpenRA.Game/World.cs-282-                                               if (ni.Actor.IsIdle)
OpenRA.Game/World.cs-283-                                                       ni.Trait.TickIdle(ni.Actor);
---
OpenRA.Game/World.cs:289:                               ActorsWithTrait<ITick>().DoTimed(x => x.Trait.Tick(x.Actor), "Trait");
--
OpenRA.Game/World.cs-299-               public void TickRender(WorldRenderer wr)
OpenRA.Game/World.cs-300-               {
OpenRA.Game/World.cs:301:                       ActorsWithTrait<ITickRender>().DoTimed(x => x.Trait.TickRender(wr, x.Actor), "Render");
OpenRA.Game/World.cs-302-               }
---
OpenRA.Game/World.cs-324-                               // hash all the traits that tick
OpenRA.Game/World.cs:325:                               foreach (var x in ActorsWithTrait<ISync>())
OpenRA.Game/World.cs-326-                                       ret += n++ * (int)(1 + x.Actor.ActorID) * Sync.CalculateSyncHash(x.Trait);
---
OpenRA.Mods.Common/AI/BaseBuilder.cs:62:                                var currentBuildings = world.ActorsWithTrait<Building>((a, b) => a.Owner == player).Count();
OpenRA.Mods.Common/AI/BaseBuilder.cs:63:                                var baseProviders = world.ActorsWithTrait<BaseProvider>((a, b) => a.Owner == player).Count();
---
OpenRA.Mods.Common/AI/BaseBuilder.cs:87:                                var currentBases = world.ActorsWithTrait<BaseProvider>((a, b) => a.Owner == player).Count();
---
OpenRA.Mods.Common/AI/BaseBuilder.cs:100:                       playerBuildings = world.ActorsWithTrait<Building>((a, b) => a.Owner == player)
OpenRA.Mods.Common/AI/BaseBuilder.cs-101-                               .Select(a => a.Actor)
OpenRA.Mods.Common/AI/BaseBuilder.cs-102-                               .ToArray();
---
OpenRA.Mods.Common/AI/BaseBuilder.cs:152:                                               cachedBuildings = world.ActorsWithTrait<Building>((a, b) => a.Owner == player).Count();
OpenRA.Mods.Common/AI/BaseBuilder.cs:153:                                               cachedBases = world.ActorsWithTrait<BaseProvider>((a, b) => a.Owner == player).Count();
---
OpenRA.Mods.Common/AI/HackyAI.cs-348-                   var myUnits = Player.World
OpenRA.Mods.Common/AI/HackyAI.cs:349:                           .ActorsWithTrait<IPositionable>((a, p) => a.Owner == Player)
OpenRA.Mods.Common/AI/HackyAI.cs-350-                           .Select(a => a.Actor.Info.Name).ToList();
---
OpenRA.Mods.Common/AI/HackyAI.cs-361-           int CountBuilding(string frac, Player owner)
OpenRA.Mods.Common/AI/HackyAI.cs-362-           {
OpenRA.Mods.Common/AI/HackyAI.cs:363:                   return World.ActorsWithTrait<Building>((a, b) => a.Owner == owner && a.Info.Name == frac).Count();
OpenRA.Mods.Common/AI/HackyAI.cs-364-           }
OpenRA.Mods.Common/AI/HackyAI.cs-365-
OpenRA.Mods.Common/AI/HackyAI.cs-366-           int CountUnits(string unit, Player owner)
OpenRA.Mods.Common/AI/HackyAI.cs-367-           {
OpenRA.Mods.Common/AI/HackyAI.cs:368:                   return World.ActorsWithTrait<IPositionable>((a, p) => a.Owner == owner && a.Info.Name == unit).Count();
OpenRA.Mods.Common/AI/HackyAI.cs-369-           }
---
OpenRA.Mods.Common/AI/HackyAI.cs:376:                   return World.ActorsWithTrait<Building>((a, b) => a.Owner == owner && Info.BuildingCommonNames[commonName].Contains(a.Info.Name))
OpenRA.Mods.Common/AI/HackyAI.cs-377-                           .Count();
---
OpenRA.Mods.Common/AI/HackyAI.cs:686:                   var newUnits = self.World.ActorsWithTrait<IPositionable>((a, p) =>
OpenRA.Mods.Common/AI/HackyAI.cs-687-                           a.Owner == Player && !a.Info.TraitInfosAny<BaseBuildingInfo>() && !activeUnits.Contains(a))
OpenRA.Mods.Common/AI/HackyAI.cs-688-                           .Select(a => a.Actor);
---
OpenRA.Mods.Common/AI/HackyAI.cs:784:                   var buildings = self.World.ActorsWithTrait<RallyPoint>((a, rp) => a.Owner == Player &&
OpenRA.Mods.Common/AI/HackyAI.cs-785-                           !IsRallyPointValid(rp.Location, a.Info.TraitInfoOrDefault<BuildingInfo>()))
OpenRA.Mods.Common/AI/HackyAI.cs-786-                           .ToArray();
---
OpenRA.Mods.Common/AI/HackyAI.cs-978-           internal IEnumerable<ProductionQueue> FindQueues(string category)
OpenRA.Mods.Common/AI/HackyAI.cs-979-           {
OpenRA.Mods.Common/AI/HackyAI.cs:980:                   return World.ActorsWithTrait<ProductionQueue>((a, q) => a.Owner == Player && q.Info.Type == category && q.Enabled)
OpenRA.Mods.Common/AI/HackyAI.cs-981-                           .Select(a => a.Trait);
OpenRA.Mods.Common/AI/HackyAI.cs-982-           }
---
OpenRA.Mods.Common/AI/States/StateBase.cs:34:                   var buildings = squad.World.ActorsWithTrait<Building>((a, b) => a.Owner == squad.Bot.Player)
OpenRA.Mods.Common/AI/States/StateBase.cs-35-                           .Select(a => a.Actor).ToList();
---
OpenRA.Mods.Common/Activities/Air/HeliReturn.cs-46-                             var rearmBuildings = heli.Info.RearmBuildings;
OpenRA.Mods.Common/Activities/Air/HeliReturn.cs:47:                             var nearestHpad = self.World.ActorsWithTrait<Reservable>((a, r) => a.Owner == self.Owner && rearmBuildings.Contains(a.Info.Name))
OpenRA.Mods.Common/Activities/Air/HeliReturn.cs-48-                                                                     .Select(a => a.Actor)
OpenRA.Mods.Common/Activities/Air/HeliReturn.cs-49-                                                                     .ClosestTo(self);
---
...
```

`ToArray` is also used by activity constructors:
```
OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:33:                      ammoPools = self.TraitsToArray<AmmoPool>();
OpenRA.Mods.Common/Activities/Air/HeliAttack.cs:49:                     ammoPools = self.TraitsToArray<AmmoPool>();
OpenRA.Mods.Common/Activities/Demolish.cs:35:                   demolishables = target.TraitsToArray<IDemolishable>();
OpenRA.Mods.Common/Activities/Move/Drag.cs:31:                  moveDisablers = self.TraitsToArray<IDisableMove>();
OpenRA.Mods.Common/Activities/Move/Move.cs:46:                  moveDisablers = self.TraitsToArray<IDisableMove>();
OpenRA.Mods.Common/Activities/Move/Move.cs:63:                  moveDisablers = self.TraitsToArray<IDisableMove>();
OpenRA.Mods.Common/Activities/Move/Move.cs:74:                  moveDisablers = self.TraitsToArray<IDisableMove>();
OpenRA.Mods.Common/Activities/Move/Move.cs:85:                  moveDisablers = self.TraitsToArray<IDisableMove>();
OpenRA.Mods.Common/Activities/Move/Move.cs:100:                 moveDisablers = self.TraitsToArray<IDisableMove>();
OpenRA.Mods.Common/Activities/Move/Move.cs:118:                 moveDisablers = self.TraitsToArray<IDisableMove>();
OpenRA.Mods.Common/Activities/Rearm.cs:25:                      ammoPools = self.TraitsToArray<AmmoPool>(p => !p.Info.SelfReloads);
OpenRA.Mods.Common/Activities/Turn.cs:25:                       moveDisablers = self.TraitsToArray<IDisableMove>();
OpenRA.Mods.RA/Activities/LayMines.cs:36:                       ammoPools = self.TraitsToArray<AmmoPool>();
```
I suppose it depends on how often activities are constructed. I know the AI can give several orders in a short amount of time, but I don't know how much is activity construction.

I'm looking at using LINQ and removing enumerators for the other stuff right now.

File: OpenRA.Game/Actor.cs (hunk 1 above)
Comment (atlimit8) [inline OpenRA.Game/Actor.cs]: Had that (`AnyTrait`), but @pchote suggested `TraitsAny` for consistency.

Comment (RoosterDragon) [inline OpenRA.Game/Actor.cs]: Fair enough.

**Comment by RoosterDragon** [comment]
I am for introducing things like ToArray and Count where you can definitely do better than the standard LINQ call because you have information it doesn't, provided the current calls are actually slow enough to warrant the extra code (I can trust the inbuilt ToArray, but I don't want to maintain a custom one unless it's actually buying me something, you know?) so I'd be down for those if profiling says it's worthwhile.

Stuff like ForEachActorHavingTrait won't ever be worthwhile performance wise though. There's so much overhead from allocating and invoking the delegates and you don't even win clarity since a plain loop would be just as easy to understand in this case.

Anyway I'm waffling a bit here. I'll try and get you some perf numbers when I get a chance. :)

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by RoosterDragon** [comment]
Once you get a chance to rebase, I'd be interested in profiling a version updated with comments so far.

In particular I've noticed the existing ```HasTrait``` is fairly allocation-heavy and it would be interesting to see if switching that to use the Infos instead makes much difference there.

**Comment by atlimit8** [comment]
Closing for now.

**Pull request closed without merging**